sql/importer: periodically check if IMPORT should replan

dt opened the pull request:
This adds a setting, bulkio.import.replan_flow_threshold, which can be set to a fraction
change in the number of instances involved in an IMPORT above which the execution of the
IMPORT will be stopped and restarted, to re-plan according to the updated cluster configuration.
The change is defined as the number of additional instances assigned work, or existing instances
assigned differnet work versus the number of instances in the existing plan. The frequency at
which this is checked can be configured with bulkio.import.replan_flow_frequency.

By default this is not enabled yet; if enabled it runs at a default frequency of 2min.
Both of these settings remain non-public for the time being.

Release note: none.

dt committed:
sql: add helper to detect physical plan changes

Release note: none.

dt committed:
sql/importer: periodically check if IMPORT should replan

This adds a setting, bulkio.import.replan_flow_threshold, which can be set to a fraction
change in the number of instances involved in an IMPORT above which the execution of the
IMPORT will be stopped and restarted, to re-plan according to the updated cluster configuration.
The change is defined as the number of additional instances assigned work, or existing instances
assigned differnet work versus the number of instances in the existing plan. The frequency at
which this is checked can be configured with bulkio.import.replan_flow_frequency.

By default this is not enabled yet; if enabled it runs at a default frequency of 2min.
Both of these settings remain non-public for the time being.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79244)
<!-- Reviewable:end -->

miretskiy commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+	initial *PhysicalPlan,
+	fn PhysicalPlanMaker,
+	execCtx JobExecContext,
+	freq time.Duration,
```
Would be nice to have frequency as a cluster setting.
If we want that to be caller controlled, we can pass `DurationSetting`

dt replied:
no reason frequency can't be cluster setting (see usage in second commit). but it doesn't seem worth passing one in and making a *much* more complicated timer that adjusts on setting change (you'd need an on-change to wake it), so it's just up to the caller to call the setting and pass it.

miretskiy replied:
But this, a long running function.  It really isn't that much more difficult to reset the timer; but we almost always wish we could change a fixed frequency.

dt replied:
It's pretty annoying/expensive to use the on-change API with a multiple waiters isn't it?

dt replied:
Switched it to take a func() time.Duration

miretskiy commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+	name string,
+	initial *PhysicalPlan,
+	fn PhysicalPlanMaker,
+	execCtx JobExecContext,
```
Does it make sense to pass DistSQLPlanner() and not expose full JobExecContext?

dt replied:
eh, JobExecContext is in the same package, and this is for use in jobs, so I don't see much benefit in adding a new interface.

miretskiy replied:
I disagree.  We should not be passing wider interfaces when narrower will do.
Furthermore, the only reason JobExecContext exists here is because of circular dependencies.

dt replied:
Eh, I suppose? Used an interface.

stevendanna commented on pkg/sql/importer/import_job.go:
```diff
@@ -1220,14 +1220,20 @@ func ingestWithRetry(
+			})
+			res, err = distImport(ctx, execCtx, job, tables, typeDescs, from, format, walltime,
+				alwaysFlushProgress, procsPerNode)
+			// Replanning errors should not count towards retry limits.
```
This seems reasonable, but do we need something to protect against 'replanning thrashing'.  Perhaps we could pass in the "last replan" time and disallow replanning too often within some time period?

dt replied:
the replan trigger owns that -- e.g. in the import change by default it won't do more than once every 2min.

dt commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+					log.Warningf(ctx, "job replanning check failed to generate plan: %v", err)
+					continue
+				}
+				changed, growth := CalculatePlanGrowth(initial, p)
```
it isn't just number of procs -- it is number of nodes which would have new _or different procs_ assigned; so even if the same three nodes have the same three procs, it should (see the todo) return 3 if you shuffled all the span assignments.

dt commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+					continue
+				}
+				changed, growth := CalculatePlanGrowth(initial, p)
+				threshold := thresholdFn()
```
I had this taking more func args at one point, but it started to feel like there was little saved by the helper if I just had all the logic repeated at the callers, with their own logging, etc. It seemed like initially "number of nodes that would do work they're not assigned now" is broadly applicable to all jobs so the helper can de-clutter the callers more by being less configurable, and taking just how often to check and how much to trip?

dt replied:
I ripped up the API a little, to make it a little more pluggable while also keeping the callers as concise as I could. WDYT?

dt commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+	execCtx JobExecContext,
+	freq time.Duration,
+	thresholdFn func() float64,
+) (func(context.Context) error, chan struct{}) {
```
Sure, yeah, we don't want anyone sending on it so that makes sense.

dt commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -95,3 +97,86 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+// CalculatePlanGrowth returns the number of new or updated sql instances in one
+// physical plan relative to an old one, as well as that number as a fraction of
+// the number in the old one.
+func CalculatePlanGrowth(before, after *PhysicalPlan) (int, float64) {
```
un-exported.

miretskiy commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -86,12 +80,104 @@ func (dsp *DistSQLPlanner) setupAllNodesPlanningTenant(
+// calculatePlanGrowth returns the number of new or updated sql instances in one
+// physical plan relative to an old one, as well as that number as a fraction of
+// the number in the old one.
+func calculatePlanGrowth(before, after *PhysicalPlan) (int, float64) {
```
I wonder if it makes sense to move this to standalone file. With test :)

dt replied:
I think this is the right file -- this file has non-query distsql planning helpers, which is what this is. 

Added test.

miretskiy replied:
ack

dt committed:
importer,sql: move plan randomization to importer

Previously sql.SetupAllNodesPlanning randomized the order of returned
sql instances, so that multiple IMPORTs that used fewer than all of the
nodes would use different sets of nodes rather than all use just the
first k nodes. This moves that randomization to import's own planning
as all other callers assign processors to specific nodes, typically by
data locality (PartitionSpans). This also allows IMPORT to randomize the
ordering using its job ID the random seed, so that within a job the order
is stable if replanned.

Release note: none.

miretskiy commented on pkg/sql/distsql_plan_bulk_test.go:
```diff
@@ -0,0 +1,82 @@
+			require.Equal(t, tc.frac, frac)
+		})
+	}
+}
```
Thank you.

miretskiy reviewed (approved):
This looks good to me.  Certainly something that can be improved on over time if needed.

dt commented:
TFTR!

bors r+

dt commented:
bors r+

dt commented:
blathers backport 22.1
